picard/util/pipe.py

```diff
@@ -0,0 +1,279 @@
+
+        # named pipe values needed by windows API
+        if self.__is_win:
+            self.__app_version = self.__app_version.replace(".", "-")
+
+            # win32pipe.CreateNamedPipe
+            # more about the arguments: https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createnamedpipea
+            self.__MAX_INSTANCES: int = 1
+            self.__BUFFER_SIZE: int = 65536
+            # timeout doesn't really matter, concurrent.futures ensures that connections are closed in declared time
+            # the value is in milliseconds
+            self.__DEFAULT_TIMEOUT: int = 300
+
+            # win32file.CreateFile
+            # more about the arguments: http://timgolden.me.uk/pywin32-docs/win32file__CreateFile_meth.html
+            self.__SHARE_MODE: int = 0
+            self.__FLAGS_AND_ATTRIBUTES: int = 0
+
+            # pywintypes.error error codes
+            # more about the error codes: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/18d8fbe8-a967-4f1c-ae50-99ca8e491d2d
+            self.__FILE_NOT_FOUND_ERROR_CODE: int = 2
+            self.__BROKEN_PIPE_ERROR_CODE: int = 109
+
+        self.path: str = self.__generate_filename()
+
+        self.is_pipe_owner: bool = False
+        self.permission_error_happened: bool = False
+
+        if self.__is_win:
+            for arg in args:
+                if not self.send_to_pipe(arg):
+                    self.is_pipe_owner = True
+                    break
+        else:
+            try:
+                self.__create_unix_pipe()
+            except FileExistsError:
+                for arg in args:
+                    if not self.send_to_pipe(arg):
+                        self.__create_unix_pipe()
+                        break
+
+    def __generate_filename(self) -> str:
+        if self.__is_win:
+            self.__pipe_parent_dir = "\\\\.\\pipe\\"
+        elif self.__is_mac:
+            self.__pipe_parent_dir = os.path.expanduser("~/Library/Application Support/MusicBrainz/Picard/pipes/")
+        else:
+            self.__pipe_parent_dir = f"{os.getenv('XDG_RUNTIME_DIR')}/"
+            # just in case the $XDG_RUNTIME_DIR is not declared, fallback dir
+            if not self.__pipe_parent_dir:
+                self.__pipe_parent_dir = os.path.expanduser("~/.config/MusicBrainz/Picard/pipes/")
+
+        return f"{self.__pipe_parent_dir}{self.__app_name}_v{self.__app_version}_pipe_file"
+
+    def __create_unix_pipe(self) -> None:
+        try:
+            try:
+                # just to be sure that there's no broken pipe left
+                os.unlink(self.path)
+            except FileNotFoundError:
+                pass
+            try:
+                os.mkfifo(self.path)
+            # no parent dirs detected, need to create them
+            except FileNotFoundError:
+                os.makedirs(self.__pipe_parent_dir, exist_ok=True)
+                os.mkfifo(self.path)
+        except PermissionError:
```

Comment:
Using a boolean here isn't that great, the actual error message isn't preserved (and it might be useful to the user).
What about raising an exception instead ?

Sopmething like:

```python
except PermissionError as exc:
   raise PipePermissionError(exc) from None
```

And in tagger code:

```python
try:
    pipe_handler = pipe.Pipe(app_name=PICARD_APP_NAME, app_version=PICARD_FANCY_VERSION_STR, args=picard_args.FILE)
    should_start = bool(pipe_handler.is_pipe_owner}
except PipePermissionError as err:
    log.error(err)
    pipe_handler = None
    should_start = True
```

Reply:
Okay for me, a log would be useful in such case.

Reply:
Also, I'll name it `PipeErrorNoPermission` to make it consistent with other errors, i.e. `f"PipeError{some_description}"`.

Also I moved `self.is_pipe_owner` to the `try` block, because it makes no sense to own a non-existent pipe

Reply:
btw, why do you call `bool()` here: `should_start = bool(pipe_handler.is_pipe_owner)`? `pipe_handler.is_pipe_owner` can  be only a boolean.

Reply:
> btw, why do you call `bool()` here: `should_start = bool(pipe_handler.is_pipe_owner)`? `pipe_handler.is_pipe_owner` can be only a boolean.

Then it's useless ;)